File: implementation/simulation/[백준]20055. 컨베이어 벨트 위의 로봇/solution.py
import sys
input = sys.stdin.readline

def rotate():
    global s, e
    s = (s - 1 + n * 2) % (n * 2)
    e = (e - 1 + n * 2) % (n * 2)
    if r[e]: r[e] = 0

def move():
    c = e
    for i in range(n - 1):
        nc = (c - 1 + n * 2) % (n * 2)
        if r[nc] and r[c] == 0 and h[c] >= 1:
            h[c] -= 1
            r[nc], r[c] = r[c], r[nc]
            if c == e: r[c] = 0 # 내리기
        c = nc

def load():
    if h[s] > 0:
        r[s] = 1
        h[s] -= 1

n, k = map(int, input().split())
h = list(map(int, input().split()))
r = [0] * n * 2
s = 0
e = n - 1
cnt = 1
while True:
    rotate()
    move()
    load()
    if h.count(0) >= k:
        print(cnt)
        break
    cnt += 1

File: implementation/simulation/[백준]20055. 컨베이어 벨트 위의 로봇/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int h[200], r[200];
int n, k, n2, s, e, cnt;

void rotate() {
    s = (s - 1 + n2) % n2;
    e = (e - 1 + n2) % n2;
    if (r[e]) r[e] = 0;
}

void move() {
    int c = e;
    for (int i = 0; i < n - 1; i++) {
        int nc = (c - 1 + n2) % n2;
        if (r[nc] && r[c] == 0 && h[c] > 0) {
            h[c] -= 1;
            if (h[c] == 0) cnt++;
            r[c] = 1;
            r[nc] = 0;
            if (c == e) r[c] = 0;
        }
        c = nc;
    }
}

void load() {
    if (h[s] > 0) {
        r[s] = 1;
        h[s] -= 1;
        if (h[s] == 0) cnt++;
    }
}

int main() {
    ios::sync_with_stdio(false); cin.tie(0);
    cin >> n >> k;
    n2 = n * 2;
    for (int i = 0; i < n2; i++) cin >> h[i];

    s = 0; e = n - 1;
    int res = 1;
    while (true) {
        rotate();
        move();
        load();
        if (cnt >= k) {
            cout << res;
            break;
        }
        res += 1;
    }
}